### DISCUSSION TITLE
[API Request] Expose additional data for media

### DISCUSSION THREAD

**Discussion by Tolriq** [opensubsonic/open-subsonic-api#23] [API requests]
### Type of change

API tweak

### Proposal description

This is a place holder request to gather all the needs by users and clients. This post will be edited with comments made.
(This is also related to https://github.com/opensubsonic/open-subsonic-api/discussions/8)

See also https://github.com/opensubsonic/open-subsonic-api/discussions/10

Add official / documented new data fields for artists / albums and songs to improve metadata.

Expose a `releaseType` String for albums when the server have the data.
Expose a `mood `String for albums when the server have the data.
Expose a `styles `String (Comma separated or multi value) for albums when the server have the data.
Expose a `bpm `Int for songs when the server have the data.
Expose a `comment `String for songs when the server have the data.
Expose a `played` String [ISO 8601] for songs when the server have the data.
Expose a `sortTitle `String for songs/albums/artists when the server have the data.
Expose a `isCompilation` Boolean for albums when the server have the data.

### Backward compatibility impact

N/A

### Backward compatibility

- [X] No backward compatibility impact.

### API details

Define the field name and type for consistency across all server.

TBD when the list is exhaustive.

**Comment by epoupon** [comment]
Just some notes:
* what do you mean by "styles"?
* we should avoid comma separated values and use multi value fields instead
* release type may not be a single field, see what is used by music brainz: https://musicbrainz.org/doc/Release_Group/Type

**Reply by Tolriq** [reply]
> what do you mean by "styles"?

This is an additional type returned by some media providers (Kodi / Plex), scraped from some online database or configured by the user. Different from genre and mood.

> we should avoid comma separated values and use multi value fields instead

Yes it's linked to #8 but just pointing here it's multi value.

 > release type may not be a single field, see what is used by musicbrainz : https://musicbrainz.org/doc/Release_Group/Type

Then maybe it should be exposed as 2 fields if necessary, currently all other providers I support (Kodi, Emby, Jellyfin and Plex expose this as a single field that can have all the values exposed from your link) and from the raw tag POV this is a single string tag.

In what tag would the subtype be exposed?

**Reply by epoupon** [reply]
> Then maybe it should be exposed as 2 fields if necessary, currently all other providers I support (Kodi, Emby, Jellyfin and Plex expose this as a single field that can have all the values exposed from your link) and from the raw tag POV this is a single string tag.
> 
> In what tag would the subtype be exposed?

Oh ok, but how do they report multiple values, like Album+Live, Compilation+Live, or even Album+Compilation+Live ?
(internally LMS uses primaryType and a set of secondaryTypes (LMS tries to stick to MB, but that's an assumed choice)).

**Reply by Tolriq** [reply]
Hum I'd need to check all one by one but for most they expose a specific isCompilation boolean. Then a single field that is raw text.
Looking at my test database from Symfonium (so from most supported providers) I have some case of the field returned as album / live and some other as just live, I have some ep and remix too returned (and of course many album or compilation one). I don't know if they expose the MB data or just the actual media tags. I can't keep up with all the internals of the servers I support :)
This is not something I personally use much (except compilation flag) so not sure what's best. All I know is that having a way to identify compilations is a must have. And having more types to filter live, soundtracks, ... also a nice to have.

**Reply by gravelld** [reply]
One issue with hard baking an attribute like compilation is that definitions of compilation differ from database to database. Some don't consider single-artist "best ofs" a compilation, only various-artists albums.

**Reply by Tolriq** [reply]
I would assume that for the vast majority of the target of Subsonic servers, people would rely on their tags to set the compilation tag. Or having the various artist. We can document what the field is for precisely, but it's a really nice information to have for clients.

**Reply by gravelld** [reply]
Yes, I guess that's only an issue if semantics are attached to the data.

**Comment by lachlan-00** [comment]
* is played a date last played?
* sorttitle is the name without a prefix? "The Mountain Goats" => "Mountain Goats"
* do many clients differentiate compilations? I don't really have any basis for them in Ampache other than something with a high song artist count maybe.

**Reply by Tolriq** [reply]
Yes it's lastPlayed but already added as played by Navidrome.
Sorttitles allows much more things via the tags, it's used a lot for artists like name Bob Dylan sort title Dylan, Bob. It's also used by many for foreign stuff with the normal name in foreign characters but the sort title in corresponding ascii text.
Yes a few apps handle compilations. It's useful for many things, including automatic display change (like you show the song artists on the song list, while on an album by a single artist it makes no sense)

**Reply by deluan** [reply]
> Yes it's lastPlayed but already added as played by Navidrome.

It was a poor naming choice, I know, but I was trying to keep consistency with `starred`, `created`, `changed`.... But I'm open to change it.

**Comment by acroyear** [comment]
Some ID3 tags I could use for improving classical music support - composer, conductor, and work.

**Reply by Tolriq** [reply]
There's another issue for those as it's tied to multi artists.

**Comment by lonebyte** [comment]
`discTitle` would be a good addition to store the disc titles of big compilations (https://musicbrainz.org/release/c44f2868-a2c2-4aa5-950b-a359357b612e/discids) or musicals (https://musicbrainz.org/release/7a3d644c-8f98-47d9-8220-1254fa849bbc/discids)

https://www.blisshq.com/music-library-management-blog/2018/03/27/lamentable-state-disc-title-support/

**Comment by Tolriq** [comment]
See proposal https://github.com/opensubsonic/open-subsonic-api/discussions/52

**Comment by certuna** [comment]
This probably fits here, but I can also split this into a new discussion: `language`. Working on that now in Navidrome but it makes more sense if it's in OpenSubsonic too.

Two considerations:

### Single or multi valued?
- [song](https://opensubsonic.netlify.app/docs/responses/song/): id3 and Vorbis allow multiple languages in the `TLAN/LANGUAGE` tags. But if you look at MusicBrainz Picard (and the Musicbrainz DB), language is stored single valued: a song can have only one language value, if there's multiple languages in the song it gets `mul` (= ISO code for multiple languages). Up to us to decide - is `language` a single or a multivalued property?
- [AlbumID3 [OS]](https://opensubsonic.netlify.app/docs/responses/albumid3/): what should servers do: roll up an album with 10 songs in `eng` and 1 song in `spa` as a single value `mul`? That really limits clients ability to search for "albums in English". Or should the API do `{eng, spa}`?
- [ArtistID3 [OS]](https://opensubsonic.netlify.app/docs/responses/artistid3/): artist with one (or more) languages associated?
### Formatting
- [ISO 639-2](https://en.wikipedia.org/wiki/ISO_639-2) (1998, as in id3v2.4) or [ISO 639-3](https://en.wikipedia.org/wiki/ISO_639-3) (2007)? Defining this field as ISO-code only makes it easy for clients to localize the language name client-side, based on system language.

**Reply by Tolriq** [reply]
This is in prod so it would require a new thread. 

As you said the standard allows multiple lang and people do use that, it also support arbitrary strings that should be supported too, the user can have a specific way to manage languages. We should not drop the value because the user set English / Spanish in the tags.

So songs should have multiple language values as array of string, albums should have the distinct list of all songs languages.

For artists since unfortunately the API does not yet support genres, I do not think we should expose language without also exposing genres that would have a lot more user interest so another request to expand artists data from song data.

For the record the usage is: `mul: Multiple languages (if you don’t use multivalue language tags)` so server should not transform multiple languages to `mul`

**Reply by certuna** [reply]
OK I'll continue in a new thread.
